negotiation and execution phase spike (#40)
comit-network/spikes: 0017-negotiation-and-execution-protocol.adoc

= How to document cnd's HTTP API =
Thomas Eizinger <thomas@coblox.tech>; Philipp Hoenisch <philipp@coblox.tech>; Daniel Karzel <daniel.karzel@coblox.tech>;
:toc:

> :revdate: 2019-10-05

Reviewer: Can we use numbered headings in here?
```suggestion
:revdate: 2019-10-05
:sectnums:
:sectnumlevels: 5
```

NOTE: Author: {authors} +
Date: {revdate} +
Tracking issue: https://github.com/comit-network/comit-rs/issues/1413[#1413]

== Context ==

> Over time we came to the conclusion the conclusion that our protocol stack in general, and RFC003 in particular is difficult to understand in certain ways.

Reviewer 2: ```suggestion
Over time we came to the conclusion that our protocol stack in general, and RFC003 in particular, is difficult to understand in certain ways.
```

We define an "offer" as part of the negotiation phase of a trade.
There can be multiple offers during the negotiation phase.
Offers are created by the maker.

=== Negotiation roles ===

* Maker: Creates offers
* Taker: Takes offers

Note that these roles are orthogonal to the cryptographic roles (Alice and Bob) in the execution protocol.
Both the maker and the taker can be in the role of Alice (who generates the secret) or Bob.
During the negotiation phase both parties have to reach agreement on who will be in the role of Alice and who will be in the role of Bob.

=== Assumptions

==== Protocols
For the sake of simplicity we assume that there is only one protocol (RFC003) for trade execution at the moment.
In the future there will be more execution protocols.

As we build upon libp2p for messaging all protocols should adhere to the libp2p specification. The proposed first negotiation protocol will be a new libp2p protocol.

==== Offers not getting stale

During the negotiation phase the offer can be changed multiple times.
Once the negotiation phase comes to a successful end (the parties agree on the parameters for the execution) the execution phase is triggered (immediately).

> Note that there is still an implicit accept of the

Reviewer: Is that what you wanted to write? 
```suggestion
Note that there might still be an implicit accept of the message sent from taker to maker with the needed parameters (e.g. similar to the _swap_request_ message right now).
```

=== Recommendation for protocol changes

> The general idea is to first define what parameters are needed for execution phase and then design a negotiation protocol that outputs these parameters for the execution.

Reviewer: ```suggestion
The general idea is to first define what parameters are needed for the execution phase and then design a negotiation protocol that outputs these parameters for the execution.
```

As RFC003 is currently the only implementation of an execution protocol we focus on this protocol.
Once more execution protocols are defined we will have to define more negotiation protocols according to the parameters required by the execution protocol.
This means, that some negotiation protocols could be re-used for multiple execution protocols if the execution protocols have the same input parameters.

==== Changes to the existing Execution protocol RFC002/RFC003

===== Input parameters

> For the execution phase we could identify the following input parameters:

Reviewer: ```suggestion
For the execution phase of RFC003 we need identify the following input parameters:
```

> 1. PeerID

Reviewer: I think it does not render because of the missing new line here.
```suggestion

1. PeerID
```

2. Ledgers (alpha and beta ledger)
3. Assets (alpha and beta asset)
4. Amounts
5. Protocol
6. Role
7. Offer ID

> 

Reviewer 2: This list does not render correctly.  Ordered list can be achieved using just the dot.

* The PeerID is needed to communicate with the CND of the counterparty.
* The offer ID is needed to link the execution of a swap back to the offer.

===== Communication changes

Instead of a request-response model we propose a messaging model where both parties supply the same parameter to the execution on both sides.
The CND nodes of the 2 parties execute the swap automatically if the parameters of the counterparty define an executable swap.

> This requires:

Reviewer: ```suggestion

This requires:
```

* Changing the messaging so that both parties can just supply the input parameters for the execution and CND sends a message to the counterparty that acknowledges.
* Logic to define if an incoming swap message is "executable"  against a created swap with matching parameters.

> 

Reviewer 2: This list does not render correctly.

==== Introducing a first Negotiation protocol

On an abstract level the negotiation protocol can be defined as a function:

```
fn negotiation(input_params[]) -> execution_params[] {

>     // steps (messages) needed to come from inptu_params to execution_params

Reviewer 2: ```suggestion
    // steps (messages) needed to come from input_params to execution_params
```

}
```

The output of the negotiation protocol are the `execution_parameters` for an execution protocol.
As we only have one execution protocol (RFC003 - Atomic Swaps using HTLCs) at the moment this negotiation protocol will be designed to output the parameters for that protocol.

TODO: More explanation.

===== Input parameters

> The minimum input parameters for the protocol are:

Reviewer: ```suggestion

The minimum input parameters for the protocol are:
```

> * Ledger/Asset pair

Reviewer: ```suggestion

* Ledger/Asset pair
```

> * PeerID (of the negotiation protocol)

Reviewer 2: Does not render.

. The Accept/Decline _feels weird_. It often does not make sense to _accept_ a trade which was already agreed on before (e.g. on application layer).
. We do not clearly distinguish rate negotiation from the execution at the moment.
. It is hard to add the concept of orders and an orderbook to COMIT.

=== This spike includes

. Solution proposal for splitting trade negotiation and execution in different protocols.
. Solution proposal for the "How to update an order"/"Orders should not get stale" problem.
. Defining the base for a first negotiation protocol.
. Defining the changes needed to the existing execution protocol RFC003.

=== This spike does not include

The needed API changes to facilitate the newly proposed execution protocol.

== TL;DR & Next step proposal

This spike proposes a new protocol to exchange trade orders between a maker and a taker including an update to the execution protocol as defined in RFC003.

We propose the following steps to achieve early validations of the proposed protocol (note: this proposal assumes that none of the assumptions have been refuted and does not make any assumptions on the size of the tickets or PRs):

> . Integrate into the https://github.com/comit-network/create-comit-app/[create-commit-app] a simple request/response example to exchange of order parameters (KISS: keep it simple stupid).

Reviewer 3: So totally discard the diagrams and just do something that works? Why?

Reviewer: What? No. Sorry if this was not clear. KISS was meant in regards of the communication protocol, i.e. don't go for libp2p or anything for now, just do a simple request/response pattern maybe using HTTP.

. Use the output of previous step and send a _swap request_ from taker to maker.
. Automate _accept swap request_ based on the defined parameters in the first few steps.
. Integrate the proposed protocol including auto-accept into the https://github.com/comit-network/comit-js-sdk[comit-js-sdk].
. Move _auto-accept_ into cnd and change cnd's API to reflect the _feeling_ of the proposed new <<execution-protocol, execution protocol>> (no actual changes of the execution protocol are needed at this point).

> . Update the proposed negotiation protocol to libp2p if needed.

Reviewer 3: But this would be the KISS protocol. I don't see a step where we actual properly define a proper negotiation protocol.

Reviewer: See above.

> . Update the communication between two cnd nodes to reflect the proposed execution protocol.

Reviewer 3: When do we review the execution protocol?

Reviewer: If you have any comments, feel free to provide them but I would expect another more clearly defined document for the execution protocol changes.

> = How to document cnd's HTTP API =

Reviewer 4: Should have a different title :sweat_smile:

Thomas Eizinger <thomas@coblox.tech>; Philipp Hoenisch <philipp@coblox.tech>; Daniel Karzel <daniel.karzel@coblox.tech>;
:toc:
:revdate: 2019-10-05
:sectnums:
:sectnumlevels: 5

NOTE: Author: {authors} +
Date: {revdate} +
Tracking issue: https://github.com/comit-network/comit-rs/issues/1413[#1413]

== Context ==

Over time we came to the conclusion that our protocol stack in general, and RFC003 in particular, is difficult to understand in certain ways.

We could specify several problems:

> . The Accept/Decline feels weird, it often does not make sense in specific use-case context

Reviewer: These dots seem to be a bit weird... shouldn't that be a `*`? 
```suggestion
* The Accept/Decline feels weird, it often does not make sense in specific use-case context
```

. We do not clearly distinguish trade negotiation from execution at the moment.
. It is hard to add the concept of orders and an order-book to COMIT.

This spike includes:

. Solution proposal for splitting trade negotiation and execution in different protocols.

> . Solution proposal for the "How to update an offer"/"Offers should not get stale" problem.

Reviewer 2: Suggest using the term 'order' since that is the term used when trading equities or forex.  Then in this context orders car be cancelled and new orders issued.

. Defining the base for a first negotiation protocol.
. Defining the changes needed to the existing execution protocol RFC003.

== Research ==

=== Constraints

Before defining the terminology certain constraints shall be defined, so we know what is already regarded as given.

==== Protocol Constraints

As we build upon libp2p for messaging all protocols should adhere to the libp2p specification. The proposed first negotiation protocol will be a new libp2p protocol.

For the sake of simplicity we assume that there is only one protocol (RFC003) for trade execution at the moment.
In the future there will be more execution protocols.

The negotiation protocol should be designed in a way that it is independent from a possible orderbook.

> It should be possible to use a central or decentral orderbook for the protocol.

Reviewer 2: ```suggestion
It should be possible to use a centralised or decentralised orderbook for the protocol.
```

==== Offers Constraints

During the negotiation phase the offer can be changed multiple times.
Once the negotiation phase comes to a successful end (the parties agree on the parameters for the execution) the execution phase is triggered (immediately).
Note that there might still be an implicit accept of the message sent from taker to maker with the needed parameters.

=== Definitions

==== Trade ====

In the COMIT context a trade is always between two digital assets.

We identify two main phases when it comes to a trade:

. Negotiation: The phase where the maker comes up with one or multiple offers.
The outcome of the negotiation phase should be the parameters needed for execution.
. Execution: The phase where the trade is executed, hence actions are executed on the ledger.

Basic idea of the steps of a trade as defined by this spike:

. Trade Invitation (maker -> world)
. Negotiation (P2P)
.. Offer Request (taker -> maker)
.. Offer (maker -> taker)
. Execution (P2P)

==== Negotiation and Execution Roles ====

The negotiation phase of a trade specifies two roles:

> * Maker: Creates offers

Reviewer: Just wording: 
```suggestion
* Maker: Makes offers
```

* Taker: Takes offers

> Note that these roles are orthogonal to the cryptographic roles (Alice and Bob) in the execution protocol.

Reviewer: Not sure if we want to have one negotiation protocol for all swap protocols, if so, I would make already a note here. We don't know yet if all protocols will have these roles. 
```suggestion
Note that these roles are orthogonal to the cryptographic roles (e.g. Alice and Bob in RFC003) in the execution protocol.
```

Both the maker and the taker can theoretically be in the role of Alice (who generates the secret) or Bob.
During the negotiation phase both parties have to reach agreement on who will be in the role of Alice and who will be in the role of Bob.

> ==== Bid and Ask ====

Reviewer 2: Suggest using the same terminology as trading forex.  'Trading pair', 'base asset', and 'quote asset'

ref: https://en.wikipedia.org/wiki/Currency_pair

Reviewer: I think we can refine this once we have a proper protocol defined. I see this spike as an experiment with a first proposal we should evaluate.

In the context of this document Bid and Ask is always from the market maker point of view.

* Bid: The amount of asset the maker offers.
* Ask: The amount of asset the maker asks for from a potential taker.

=== Recommendation for protocol changes

The general idea is to first define what parameters are needed for the execution phase and then design a negotiation protocol that outputs these parameters for the execution.

As RFC003 is currently the only implementation of an execution protocol we focus on this protocol.
Once more execution protocols are defined we will have to define more negotiation protocols according to the parameters required by the execution protocol.
This means, that some negotiation protocols could be re-used for multiple execution protocols if the execution protocols have the same input parameters.

=== Invitation phase

We introduce an invitation phase in order to enable decentralization.
Makers don't necessarily have a central order book, the invitation helps them to overcome the problem of decentral offers getting stale.

On an abstract level an invitation protocol can be defined as a function:

```

> fn invitation() -> negotiation_params[] {

Reviewer 3: I am not clear on that. Maybe define what `negotiation_params`. Are those the parameters to start the negotiation? Also, I'd prefer `NegotiationParams` as it's a type.

Reviewer: See below. I'll update the description and add a reference.

    // steps (messages) needed to come from nothing to negotiation_params
}
```

==== Communication

Message Overview:

> ![Invitation](http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/comit-network/spikes/master/assets/0017-SequenceDiagram-InvitationPhase.puml&fmt=svg)

Reviewer 3: The diagram is just an hello world, is that expected?
Hum, ok I see, it's in the diagram shared on Slack.

Author: the diagram shared on slack does NOT match the diagrams in the spike directly. The diagram on slack is more for thinking about actual products, whereas the spike is more "theoretical" 😃

==== Trade Invitation ====

The trade invitation specifies which assets on which ledgers a maker _is willing_ to trade.
The trade invitation is not necessarily part of the negotiation.
Its purpose is to connect maker and taker.

An asset-availability is defined by:

[%header,cols=1*]
|===
|required

|connection-info
|bid-asset
|ask-asset
|===

[%header,cols=1*]
|===

> |optional

Reviewer: I would see an optional `id` here as well. This allows tracking posted offers. 
Think of: I post on Twitter and send an offer "speciale" to Anna personally. I want to distinguish the incoming requests so that Anna gets the special offer and not everyone on twitter. 
Once negotiated, this ID can be dropped, i.e. it is different to the ID mentioned below.

|timestamp
|bid-amount
|ask-amount
|===

Constraints:

* Connection-info currently consists of IP-address and libp2p peer-id
* The optional parameters help takers to understand what a maker was capable of at a certain point in time.

Representation in JSON:

```json
{
    "connection-info": { ... },
    "bid": {
        "ledger": ...
        "asset": ...
    },
    "ask": {
        "ledger": ...
        "asset": ...
    }
}
```
When designing the negotiation protocol

> One should aim for a standardized format for trade invitations to be able to process them in the negotiation phase.

Reviewer: Any idea of how to get rid of `ledger` and `asset` combination in this step? 
I'd say that ticker symbols these days are unique and distinguishable and hence can be used for this. E.g. BTC, ETH, PAY, WBTC (Wrapped BTC on ETH), ...

One exception I was able to find: USDT, seems like there is no difference in USDT on Ethereum, Tron or Omni (Bitcoin).

Reviewer 2: I was going to say just trading pair of assets.  Assets have a symbol (BTC), a name/description (bitcoin), and with @bonomat's comment a ledger (perhaps this can be empty for symbols that exist on a single ledger only.

=== Negotiation protocol

On an abstract level the negotiation protocol can be defined as a function:

```

> fn negotiation(input_params[]) -> execution_params[] {

Reviewer 3: ```suggestion
fn negotiation(negotiation_params[]) -> execution_params[] {
```
I assume?

    // steps (messages) needed to come from input_params to execution_params
}
```

The output of the negotiation protocol should enable the execution phase.
As we only have one execution protocol (RFC003 - Atomic Swaps using HTLCs) at the moment this negotiation protocol will be designed to output the parameters for that protocol.

Asset-availability handling (i.e. building a central order-book) is not part of this protocol.

The invitation phase may be part of an implementation of a negotiation protocol.

==== Negotiation Input Parameters

The minimum input parameters for the protocol are:

* ledger/asset pair
* connection-info (of the negotiation protocol)

Example of one input parameter as JSON object:
```
{
    "bid_ledger": ... ,
    "bid_asset": ... ,
    "ask_ledger": ... ,
    "ask_asset": ... ,

>     "peer_id" ...

Reviewer 3: What about the `order id` or `swap id`. This needs to be part of the negotiation. 

If it is not, then during the execution, how can the maker tell the taker "param X did not match" if it does not know (for sure) what offer we are talking about?

}
```

These input parameters are the output of the invitation phase.

==== Negotiation Output Parameters

The output of the negotiation protocol should enable the execution phase.
However, the negotiation phase should be functionally decoupled from the execution phase.

The execution phase needs the following parameters on both sides in order to start the execution:

* role
* ledger-asset pairs
* amounts
* execution protocol
* expiries
* identities

Out of these parameters the following are relevant for negotiating the price:

* ledger-asset pairs
* amounts
* execution protocol
* expiries

Constraints:

> * Execution protocol and expiries CAN have an implication on the price.

Reviewer 3: ```suggestion
* Execution protocol and expiries MAY have an implication on the price.
```

`CAN` is not a valid word of RFC2119 so it is strange to use it and cap it. 
http://www.faqs.org/rfcs/rfc2119.html

> Thus, they should be considered as part of the negotiation.

Reviewer 3: ```suggestion
Thus, they SHOULD be considered as part of the negotiation.
```

It'd be good to have a rewrite upper-casing RFC2119 keywords.

If the expiries are not taken into account during the negotiation the execution phase should be able to come up with default expiry times.

> * Identities don't have an implication on the price negotiation and should not be exposed to the negotiation protocol.

Reviewer: I like that.

|ask-asset
|ask-amount
|===

===== P2P Offer

Message from the maker to the taker that specifies a specific offer to be executed.

[%header,cols=1*]
|===
|required

|offer id
|valid until
|bid-ledger
|bid-asset
|bid-amount
|ask-ledger
|ask-asset
|ask-amount
|_execution_ protocol(s)
|_execution_ connection-info
|===

[%header,cols=1*]
|===
|optional

|_execution_ expiries
|===

Constraints:

* The offer ID is a unique identifier for one specific offer.
* Once the taker send an offer request, the returned offer by the maker is for the taker specifically.
It has a timestamp defining until when it is valid.

The offer ID is used to be able to link the swap execution back to the negotiation.

```json
{
    id: "unique_identifier"
    "connection-info": { ... },
    "bid": {
        "ledger": ...
        "asset": ...
    },
    "ask": {
        "ledger": ...
        "asset": ...
    }
}
```

===== P2P Take Offer

Message from the taker to the maker that specifies that the taker wants to take a specific offer previously received from the maker.

[%header,cols=1*]
|===
|required

|offer id
|===

[%header,cols=1*]
|===
|optional

|_execution_ expiries
|_execution_ protocol
|===

> The taker can suggest expiries when taking the offer.

Reviewer 3: If the taker can suggest expiries, how does the maker decline them?

This is meant for use cases, where the maker did not suggest expiries in the offer, but the taker does not want to fallback to the default expiries of the execution.

The execution protocol may be specified by the taker if the maker specified more than one in his offer.

==== Shortcomings

> This first simple negotiation protocol does not really specify an orderbook in the sense that a user can compare ALL available offers.

Reviewer 3: This sounds out of scope to me as at no point was the transport mode for the invitation specified. Only the negotiation protocol is restricted to libp2p.

Because we do not specify the invitation mode then a maker could publish invitations on a  centralised order book, negating this shortcoming.

A taker can only process the offers of known makers.
Every taker is responsible for keeping its own list of known makers and evaluate the offers.
There may exist unknown makers that would offer a better rate.

=== Changes to the existing execution protocol RFC002/RFC003

Constraints:

. Taker defaults to Alice, Maker defaults to Bob
. The taker has received connection information to contact the maker for execution in the negotiation phase.

==== Input parameters

===== Execution Parameter Definition

For the execution phase as specified by RFC003 we need the parameters for creating the HTLCs on both sides:

* alpha-ledger
* alpha-asset
* alpha-amount
* beta-ledger
* beta-asset
* beta-amount
* expiries
* hash(secret)
* identities

In order to come up with these parameters we need:

> * role (to know who comes up with the secret)

Reviewer 3: We just said that the maker SHOULD be Bob so this requirement seems low priority.

In order to enable more protocols in the future we need:

* protocol

===== Negotiation Output

From the negotiation phase we have the following output parameters:

[%header,cols=1*]
|===
|required

|offer id
|valid until
|bid-ledger
|bid-asset
|bid-amount
|ask-ledger
|ask-asset
|ask-amount
|_execution_ protocol
|_execution_ connection-info (know only to taker)
|_execution_ role
|===

[%header,cols=1*]
|===
|optional

|_execution_ expiries
|===

The taker has received the execution connection information from the maker during the negotiation phase. Thus, the taker has to make the first move.

===== Parameter Mapping

> This section defines the input parameters for cnd and

Reviewer 3: The sentence is not finished.

[%header,cols=2*]
|===
|required (cnd)
|negotiation

|offer id
|offer id

|beta-ledger
|bid-ledger

|beta-asset
|bid-asset

|beta-amount
|bid-amount

|alpha-ledger
|ask-ledger

|alpha-asset
|ask-asset

|alpha-amount
|ask-amount

|protocol
|_execution_ protocol

| connection-info
|_execution_ connection-info

|role
|_execution_ role

|===

[%header,cols=2*]
|===
|optional (cnd)
|negotiation

|expiries
|_execution_ expiries

> |identities

Reviewer 3: How is it optional? Both parties need the addresses of both parties to calculate the HTLCs.

Author: It is optional from the input parameter point of view, because we are opting for `cnd` being able to come up with identities if they are not provided from the outside. 

For expiries we would have a calculation of default expiries according to a formula. 

For identities cnd could e.g. generate addresses from a transient key.

|
|===

Constraints:

* The connection-info is only provided to the cnd in the role of Alice.
* If the expiries were not part of the negotiation output cnd will come up with defaults.

> * Identities might be provided by the user. If no identities are provided the cnd will come up with defaults.

Reviewer 3: How can one party come up with a default identity for another party?

==== Communication changes

> Instead of a request-response model we propose a messaging model where both parties supply the same parameter to the execution on both sides.

Reviewer 3: Please clarify what you mean with "supply".
Do you mean the Maker App supply to the Maker cnd the same parameters than the Taker App supplies to the Taker cnd?

> The CND nodes of the 2 parties execute the swap automatically if the parameters of the counterparty define an executable swap.

Reviewer 3: ```suggestion
The cnd nodes of the 2 parties execute the swap automatically if the parameters of the counterparty define an executable swap.
```

Reviewer 3: What are "the parameters of the counterparty"?

Author: Both create the swap on their side according to what they agreed on in the negotiation spike. Then both cnd nodes "exchange messages" to verify that the parameters match. If the parameters match the execution will start automatically. 

(more constraints described in comment below https://github.com/comit-network/spikes/pull/40#discussion_r338890315)

> Thus, once one party adds a swap to the pool, cnd sends a message to the counterparty's cnd to verify that the same swap is defined there too.

Reviewer 3: What if it is not? Chicken and Egg problem here.
1. Taker informs its cnd
2. Taker cnd asks Maker cnd if same swap is defined
3. Maker cnd says no
4.??? Does the Taker cnd drop the swap?
5. Maker informs its cnd
6. Maker cnd remembers what Taker cnd asks?
7. Maker cnd inform Taker cnd. meaning we always have 2 round of messages?

Author: I discussed this with @thomaseizinger quite lengthy. Eventually we might have to opt for a more "open" messaging mode, where both the maker and the taker can send messages to request stuff from the other party.

For now I would propose that:
1. Taker cnd starts by sending all parameters to maker
2. Maker cnd evaluates if parameters match
  a. Parameters match -> send identities back
  b. Parameters don't match
    1. inform the Taker what is missing
    2. the taker can send another match request.

Constraints: 
- The maker dictates, the taker has to follow (or just abort)
- There is a timeout defined on the maker side that would automatically remove the trade if there is no match within X time.

Reviewer 3: Ok, please clarify that in the document that the taker is the one to send parameters first to confirm.
Which means
Taker:
	1. App tells cnd: here are some parameters
	2. Taker cnd sends message to Maker cnd to confirm swap
	3. Maker cnd waits time X for Maker app to add same swap id (if not done)
	4. Maker cnd replies to Taker cnd

Maker:
	1. App tells cnd: here are some parameters
	2.a (taker cnd already contacted maker cnd): Maker cnd replies to taker cnd to confirm swap
	2.b (taker cnd did not yet contact maker cnd): Maker cnd does nothing and wait for a message from taker cnd.

This requires:

> * Changing the messaging so that both parties can just supply the input parameters for the execution and CND sends a message to the counterparty to compare the parameters.

Reviewer 3: ```suggestion
* Changing the messaging so that both parties can just supply the input parameters for the execution and cnd sends a message to the counterparty to compare the parameters.
```

** The swap is deemed executable once the parameters match on both sides.
* Once the parameters match the taker (which defaulted to Alice) gets the fund transaction, the maker cnd monitors alpha-ledger for funding.

Messaging:

![Execution](http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/comit-network/spikes/master/assets/0017-SequenceDiagram-ExecutionPhase.puml&fmt=svg)

== Summary of recommendations ==

=== Optimizing for the user

In the "best case" scenario, i.e. a scenario without any errors that assumes cnd coming up with default identities and default expiries the human interaction is kept to a bare minimum:

. Maker: create and publish trade invitation
. Taker: take offer
. Taker: fund
. Maker: fund
. Taker: redeem
. Maker: redeem

Constraints:

. The actual offer (in the maker app) can be kept up to date automatically on the maker side.
. Offer invitations can be automatically processed on the taker side.
. Offer requests can be automated sent (periodically query for the current rate) on the taker side.
. The execution phase assumes two cnd that run compatible versions.

> . In order to come up with indentities cnd has to use tansient keys.

Reviewer: ```suggestion
. In order to come up with identities cnd has to use transient keys.
```
